llmfoundry/models/layers/ffn.py

```diff
@@ -23,34 +29,58 @@ def __init__(
         d_model: int,
         expansion_ratio: int,
         fc_type: str = 'torch',
+        ffn_hidden_size: Optional[int] = None,
         device: Optional[str] = None,
         bias: bool = True,
     ):
         super().__init__()
-        fc_kwargs: dict[str, Any] = {
+        if ffn_hidden_size is not None:
+            log.info(ignore_expr_log_msg % (expansion_ratio, ffn_hidden_size))
+        else:
+            ffn_hidden_size = d_model * expansion_ratio
+            if ffn_hidden_size != int(ffn_hidden_size):
+                raise ValueError(
+                    f'`d_model * expansion_ratio` ({ffn_hidden_size}) must be an integer.'
+                )
+            ffn_hidden_size = int(ffn_hidden_size)
+
+        self.fc_kwargs: dict[str, Any] = {
             'bias': bias,
         }
         if fc_type != 'te':
-            fc_kwargs['device'] = device
+            self.fc_kwargs['device'] = device
+
         self.up_proj = FC_CLASS_REGISTRY[fc_type](
             d_model,
-            expansion_ratio * d_model,
-            **fc_kwargs,
+            ffn_hidden_size,
+            **self.fc_kwargs,
         )
         self.act = nn.GELU(approximate='none')
         self.down_proj = FC_CLASS_REGISTRY[fc_type](
-            expansion_ratio * d_model,
+            ffn_hidden_size,
             d_model,
-            **fc_kwargs,
+            **self.fc_kwargs,
         )
         self.down_proj._is_residual = True
 
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         return self.down_proj(self.act(self.up_proj(x)))
 
 
+class MPTGeGLU(MPTMLP):
+
+    def __init__(self, **kwargs: Any):
+        super().__init__(**kwargs)
+        self.gate = deepcopy(self.up_proj)
```

Comment:
Any reason you're not just constructing a new layer here and instead doing a deepcopy? I find it a bit unintuitive with the deepcopy, because I have to think about if there are any weird interactions resulting from copying.

Reply:
Just trying make as few line changes as possible (for no good reason), but I can make a new layer